src/util/segmentation/operations/correction.js

```diff
@@ -0,0 +1,524 @@
+import { fillInside, eraseInside } from './index.js';
+import getPixelPathBetweenPixels from './getPixelPathBetweenPixels';
+import clip from '../../clip';
+import pointInImage from '../../pointInImage';
+
+import { getLogger } from '../../logger';
+import floodFill from './floodFill.js';
+
+const logger = getLogger('util:segmentation:operations:correction');
+
+/**
+ * correction - Using the stroke given, determine which action(s) to perfom:
+ * - Stroke starts and ends inside a segmentation: Behaves as an subtractive freehand scissors.
+ * - Stroke starts and ends outside a segmentation: Behaves as an additive freehand scissors.
+ * - Stroke out-in-out: Section is subtracted.
+ * - Stroke in-out-in: Section is added.
+ *
+ * @param  {Object} evt The cornerstone event.
+ * @returns {null}
+ */
+export default function correction(evt) {
+  const { operationData } = evt;
+
+  if (operationData.segmentationMixinType !== `freehandSegmentationMixin`) {
+    logger.error(
+      `correction operation requires freehandSegmentationMixin operationData, recieved ${
+        operationData.segmentationMixinType
+      }`
+    );
+
+    return;
+  }
+
+  const { pixelData, segmentIndex } = operationData;
+
+  const nodes = snapPointsToGrid(evt);
+
+  if (simpleScissorOperation(nodes, evt, segmentIndex)) {
+    return;
+  }
+
+  const operations = splitLineIntoSeperateOperations(nodes, segmentIndex);
+
+  // Create binary labelmap with only this segment for calculations of each operation.
+  const workingLabelMap = new Uint8Array(pixelData.length);
+
+  operations.forEach(operation => {
+    performOperation(operation, pixelData, workingLabelMap, segmentIndex, evt);
+  });
```

Comment:
Mostly a style thing, but:

```js
 // Create binary labelmap with only this segment for calculations of each operation.
 const workingLabelMap = new Uint8Array(pixelData.length);
 const operations = splitLineIntoSeperateOperations(nodes, segmentIndex);

  operations.forEach(operation => {
    performOperation(operation, pixelData, workingLabelMap, segmentIndex, evt);
  });
```

Group const declarations? Not sure if there's a rule we can create for it, but it makes me happy.

Reply:
Is `performOperation` doing something other than `fillInside` / `fillOutside` on a subset of nodes? It looks like for our complex case, we do a lot more magic than I would have guessed.

Reply:
Its not a `fillInside` or `fillOutside` at all, as what gets filled depends not only on the set of nodes, but also on the geometry of the segment current present on the labelmap:

![img](https://media.giphy.com/media/mDZ9mBMwDJqiq4GF5X/giphy.gif)

Reply:
I appreciate the explanation. Thank you.